coreax/coresubset.py

```diff
@@ -617,3 +620,294 @@ def _loop_body(
+        try:
+            # Note that a TypeError is raised if the size input to jnp.zeros is negative
+            coreset_indices = jnp.zeros(coreset_size, dtype=jnp.int32)
+
+        except TypeError as exception:
+            if coreset_size < 0:
+                raise ValueError("coreset_size must not be negative") from exception
+            if isinstance(coreset_size, float):
+                raise ValueError(
+                    "coreset_size must be a positive integer"
+                ) from exception
+            raise
+
+        # Evaluate the diagonal of the Gram matrix
+        stein_kernel_diagonal = vmap(
+            self.stein_kernel.compute_elementwise, in_axes=(0, 0), out_axes=0
+        )(
+            self.original_data.pre_coreset_array,
+            self.original_data.pre_coreset_array,
+        )
+
+        regularised_log_pdf = 0.0
+
+        if self.regularise:
+            # Compute positive Laplace operator
+            def laplace_positive(x_):
+                hessian = jacfwd(score_function)(x_)
+                return jnp.clip(jnp.diag(hessian), a_min=0.0).sum()
+
+            laplace_positive_vector = vmap(laplace_positive, in_axes=(0,), out_axes=0)(
+                self.original_data.pre_coreset_array
+            )
+            # Add to kernel diagonal
+            stein_kernel_diagonal += laplace_positive_vector
+
+            # Fit a KDE to estimate log PDF: note the transpose
+            kde = gaussian_kde(
+                self.original_data.pre_coreset_array.T,
+                bw_method=float(self.kernel.length_scale),
+            )
+
+            # Use suggested regularisation parameter
+            regulariser_lambda = 1 / coreset_size
+            regularised_log_pdf = regulariser_lambda * kde.logpdf(
+                self.original_data.pre_coreset_array.T
+            )
+
+        # Greedily select coreset points
+        body = partial(
+            self._greedy_body,
+            x=self.original_data.pre_coreset_array,
+            kernel_vectorised=self.stein_kernel.compute,
+            kernel_diagonal=stein_kernel_diagonal,
+            regularised_log_pdf=regularised_log_pdf,
+            unique=self.unique,
+        )
+        try:
+            coreset_indices, kernel_similarity_penalty = lax.fori_loop(
+                lower=0,
+                upper=coreset_size,
+                body_fun=body,
+                init_val=(coreset_indices, kernel_similarity_penalty),
+            )
+        except IndexError as exception:
+            if coreset_size == 0:
+                raise ValueError("coreset_size must be non-zero") from exception
+            raise
+
+        # Assign coreset indices & coreset to original data object
+        self.coreset_indices = coreset_indices
+        self.coreset = self.original_data.pre_coreset_array[self.coreset_indices, :]
+
+    @staticmethod
+    @partial(jit, static_argnames=["kernel_vectorised", "unique"])
+    def _greedy_body(
```

Comment:
This function repeats a lot of what is in the `_greedy_body` of `KernelHerding`. We can reduce the duplication by having `_greedy_body(..., selection_function: Callable[[Array], Array])`, where  `index_to_include_in_coreset = selection_function(current_kernel_similarity_penalty)`. This would require `_greedy_body` to become a module level function.

Reply:
I agree. I haven't implemented it here as I think this would need to be a separate ticket.

Reply:
We can lump this into the #581 refactor